rfd/0178-app-custom-icons.md

```diff
@@ -0,0 +1,97 @@
+---
+author: Lisa Kim (lisa@goteleport.com)
+state: draft
+---
+
+# RFD 178 - Application Resource Custom Icon
+
+## Required Approvers
+
+- Engineering: @romant && @avatus && (someone from the scale team))
+
+## What
+
+Support custom icon uploads for resources through yaml. We can start with feature request [#4712](https://github.com/gravitational/teleport/issues/4712) where
+users provide a URL and Teleport's auth server will download the image and store it in the backend.
+
+## Why
+
+Users want to be able to recognize their applications by famaliar icons/logos rather than reading the application name.
+
+Today, the web UI has some hard coded methods to [guess](https://github.com/gravitational/teleport/blob/dc1199636558f85ed275512175a12207a8034d44/web/packages/shared/components/UnifiedResources/shared/viewItemsFactory.ts#L184) which icon to show for an application by looking at the application name (which may or may not be an accurate depiction of the application) and if nothing matches, we use a generic icon to mean `application`. The guess type is limited to a few handpicked applications.
+
+## Details
+
+### User Story
+
+Alice is a system admin and wants to add custom icon for application resources. Alice specifies the icon URL field on the application resource yaml and using `tctl` creates or updates this file.
+
+### Requirements
+
+- Persist icons in a storage
+- Efficiently retrieve icons
+- Upload icons via yaml resource
+
+### Implementation
+
+#### Persist icons in a storage
+
+Icons can be stored in the backend as blobs with its own row with the key formatted as `icon/<resource-kind>/<name-of-resource>:<uid>`. A unique identifier is required for the proxy cache, otherwise updated icons will not be reflected right away. The UID will be saved in the resource's spec.
```

Comment:
- This scheme allows multiple icons to be stored for the same resource. Is that something we want to support?
- We'll also need to validate that the icon does not exceed the size of a single backend item (400KB for DynamoDB, IIRC)